internal/expr/parser.go

```diff
@@ -129,25 +129,81 @@ func (p *Parser) Parse(input string) (expr *ParsedExpr, err error) {
 			continue
 		}
 
-		if sp, ok, err := p.parseStringLiteral(); err != nil {
-			return nil, err
-		} else if ok {
-			p.add(sp)
-			continue
-		}
-
 		if p.pos == len(p.input) {
 			break
 		}
 
-		// If nothing above can be parsed we advance the parser.
-		p.pos++
+		// If nothing above can be parsed we advance the parser to the next
+		// character of interest.
+		if err := p.advance(); err != nil {
+			return nil, err
+		}
 	}
 	// Add any remaining unparsed string input to the parser.
 	p.add(nil)
 	return &ParsedExpr{p.parts}, nil
 }
 
+// advance increments p.pos until it reaches content that might preceed a token
+// we want to parse.
+func (p *Parser) advance() error {
+
+	// The byte following these bytes might be the start of an expression.
+	delimiterBytes := map[byte]bool{
+		' ':  true,
+		'\t': true,
+		'\n': true,
+		'\r': true,
+		')':  true,
+		'(':  true,
+		';':  true,
+		'=':  true,
+		',':  true,
+	}
+
+	for p.pos < len(p.input) {
+		if ok, err := p.skipStringLiteral(); err != nil {
+			return err
+		} else if ok {
+			continue
+		}
+		p.pos++
+		if p.pos < len(p.input) && delimiterBytes[p.input[p.pos]] {
+			break
+		}
+	}
+
+	p.skipBlanks()
+
+	return nil
+
+}
+
+// skipStringLiteral jumps over single and double quoted sections of input.
+// Doubled up quotes are escaped.
+func (p *Parser) skipStringLiteral() (bool, error) {
+	mark := p.pos
+	if p.skipByte('"') || p.skipByte('\'') {
+		c := p.input[p.pos-1]
+
+		// We keep track of whether the next quote has been previously
+		// escaped. If not, it might be a closer.
```

Comment:
Might be easier to understand, if we change the comment to: 
"If not, it might a closing quote."